### ISSUE TITLE
icing effect using multiple dirt texture

### ISSUE THREAD

**Issue by HHS81** [HHS81/c182s#144]
When icing effect implemented use multiple dirt map  in the shader for

**Comment by HHS81** [comment]
What happend to this C182S?

- [ ] Obviously it flew through clouds of canned whipped cream
- [ ] Doc E. Brown used it - "_The way I see it, if you're gonna build a time machine into a aircraft, why not do it with some *style?*_ "
- [ ] attack of white flys

or

- [ ] the experimental icing effect?

![fgfs-screen-005](https://user-images.githubusercontent.com/12649375/33808780-33df355e-ddec-11e7-8371-4e03911bbfa9.jpg)

**Comment by gilbertohasnofb** [comment]
This looks amazing!

**Comment by hbeni** [comment]
Is it already influencing the flight dynamics and what can i do to de-ice? (Click "repair"?)
I cant wait to try the effect by myself. What causes icing?

**Comment by HHS81** [comment]
Not yet

Currently by putting it into the sun (Flights into known icing condition are forbidden with this Type of aircraft.) I'm not sure if I should add ground deicing equipement, as small aircraft are not deiced like bigger aircraft (Flights into  known icing condition are not allowed).

Cause of icing: (simplified) Freezing temperatures and high humidity.

**Comment by HHS81** [comment]
I use the already existing Code from the Seneca II,    works quite good, but need a bit to be extended to be more realistic.

**Comment by hbeni** [comment]
It would be cool to have an option to manually remove frost and ice. The POH states this in the preflight i think.

**Comment by wlbragg** [comment]
Really looks good. This would be a nice system for the heavy aircraft to have (when ready, someone might want to ping the forum with it). My thoughts on it for these light aircraft, if what you say is true about not allowed to fly into/during icing conditions, would be it is only useful for simulating a startup condition after sitting in freezing weather. Unless your interest in it is to model what would happen if you were to try to take off and fly in this condition. It certainly would add some realism to the conditions your trying to fly in.

**Comment by wlbragg** [comment]
>My thoughts on it for these light aircraft, if what you say is true about not allowed to fly into/during icing conditions, would be it is only useful for simulating a startup condition after sitting in freezing weather. 

Then again we model frost and fog on the windshield when in theory we shouldn't be flying when that condition is possible, at least the frost portion.

**Comment by HHS81** [comment]
It is a bit more complex. I will write later about.

**Comment by HHS81** [comment]
@hbeni @gilbertohasnofb @wlbragg 
### First part: Icing in realitity
Icing is a very dangerous situation for aircraft, especially for those smaller General Aircraft, which are mostly not being equiped with Anti-Icing Equipement. The Cessna 182 is not equiped with (beside the usual pitot-heater), and so the FAA did not approved it for flight into known icing conditions (= FIKI).

Ice disrupts the airflow, and so it increases drag, decreases lift, and add some small amount of extra weight. Ice accumulates on every exposed surface like the front, wings, struts, landing gear, vents, antennas etc... So it also can break antennas due to vibrations, blocking the air intake for the engine or jam the control surfaces,  and with the extra drag and less lift let the aircraft stall much faster. Short said: a completly iced aircraft can't continue it's flight. There had been a high number of deadly accidents due to icing, from small GA to bigger airliners.

There are three types of ice: Rime (milky white); clear (hard to break) and mixed (mix of both types)
Icing appears at temperatures lower than 0°C and air mass containing liquid water. Clouds can contain a lot of liquid water, even at Temperature lower than 0°C. Those super-cooled water droplets will form to ice when hitting objects like aircraft surfaces. A lot of factors play a role when it comes to icing, but up to a certain point icing condition can be predicted today, so the pilot can plan their flight to prevent icing. Especially weather fronts, cold pressure areas and freezing rain contain high risk for icing. 

In flight icing is described as:
**Trace:** Ice becomes perceptible. Rate of accumulation is slightly greater than the rate of sublimation. 
**Light:** The rate of accumulation may create a problem if flight is prolonged in this environment (over
one hour). 
**Moderate:** The rate of accumulation is such that evenshort encounters become potentially hazardous anduse of deicing/anti-icing equipment or flight diversion is necessary. 
**Severe:** The rate of accumulation is such that deicing/anti-icing equipment fails to reduce or control
the hazard. An immediate flight diversion is necessary. 

How to prevent icing condition: 
-keep of clouds, or keep your in time inside clouds short
-when icing formed on the wings etc. climb or descent to warmer air/ cloud-free altitude.

More: 
[https://www.aopa.org/-/media/Files/AOPA/Home/Pilot-Resources/ASI/Safety-Advisors/sa11.pdf](https://www.aopa.org/-/media/Files/AOPA/Home/Pilot-Resources/ASI/Safety-Advisors/sa11.pdf)
[https://www.skybrary.aero/index.php/In-Flight_Icing](https://www.skybrary.aero/index.php/In-Flight_Icing)

**Comment by HHS81** [comment]
### Second part: icing in FlightGear
Until now, the Seneca II is the only aircraft in FGAddon which simulates icing with an effect on the flight dynamic. It has a realistic good icing model, where icing will ocur if: 
-temp <0°C and effective-visibility <1000m
It also simulates the rate of accumulation depending on temperature and visibility, so you will have different rate of accumulation like trace, light moderate, Severe or Sublimation. You can also specify the objects like wings, tail, etc. which are influenced by icing, and set their individual sensibility for icing.
And since the real SenecaII is equiped with Anti-Icing Equipement, you can also specify which obects are equipped with Anti-Ice.
 
The icing model is based on the former weather system (pre-AW) in older FlightGear versions where FlightGear detected the position of the clouds and reduced automatically the effective visibility when the aircraft is in the cloud. 
With the todays weather system (default and AW) clouds position can't be detected exactly anymore, and effective-visibility is only reduced with active precipitation or fog, not automatically inside the cloud.

FlightGears now Weather System is really good, but it doesn't really simulate clouds with different amounts of liquid water, so there is no exact prediction of icing condition possible. 

That means, flying with the SenecaII will give us less ice than in reality.

**Comment by HHS81** [comment]
In reality you can deice a small aircraft like the C182 with deicing-fluid, and must remove the ice and frost when the aircraft had been parked outside and ice had been formed. You can fly as long you stay away from icing condition (cold air with high humidity).

In FlightGear we only get the current weather, so we don't know what happened hours before, and don't know if the aircraft had been covered with ice in this time. 
In FlightGear icing occurs as soon icing conditions are there - deicing could be simulated but realistically it would give us only some short time until icing start again, preventing a successful flight.

@wlbragg  frost and fog on the windshield appears **inside** the aircraft, due the moisture of exhaled air inside the aircraft cabin. So you can have sunny weather, no clouds, freezing temperatures and with that no icing condition; but still frost and fog inside the cabin on non-heated surfaces like the windshield.

@hbeni  any aircraft should be cleaned outside at the end of a flight day. So I will add a "Clean Aircraft button" which will remove the ice. I don't think I will add a "deice fluid function", which will prevent icing, since it might give unexperienced pilots a false impression since the POH clearly states on page 2-8 "Limitations" that flights into known icing conditions is prohibited.

**Comment by hbeni** [comment]
Sounds good! Should i add an item to the preflight checklist so ingame pilots get a reminder?

**Comment by hbeni** [comment]
> So I will add a "Clean Aircraft button" which will remove the ice.

I think we should add that _also_ to the "repair" button, what do you think?

>I don't think I will add a "deice fluid function"

I agree 100%

**Comment by HHS81** [comment]
I don't think we need any extra item for cleaning the aircraft yet. I thought as an idea to make the aircraft dirty by flighttime, but this isn't ready yet. And I don't know how much perfomance penalty a dirty aircraft has. usually it is good airmenship to clean your aircraft at the end of the day.

**Comment by HHS81** [comment]
I think we should put "repair" and "clean" under the same item then.

**Comment by HHS81** [comment]
I have implemented the in-cloud-icing. It will detect the vertical layer of clouds, which clouds, and check if the aircraft is inside. If inside it will trigger a signal to create ice, based on Torsten Dreyer's solution from the SenecaII.

I also made the fdm react to drag and loss of lift due to ice. But it needs some tuning, the effects seems to small.
What's missing is the effect of prop icing and increase of weight by ice. 

There are limitations: 

- we can't detect the horizontal position of clouds, and we can't detect the exact altitude of cloud base and cloud top, it is just an approximation by observation in different weather. 
- At the moment we can't detect (yet?) thunderstorm clouds, though their position are known(`/local-weather/clouds/tile[1]/cloud[...]`) - but how to read out and check 100 clouds positions?

**Comment by hbeni** [comment]
> I have implemented the in-cloud-icing.  [...] I also made the fdm react to drag and loss of lift due to ice

Very nice! Also that the pitot now blocks is very cool!

> At the moment we can't detect (yet?) thunderstorm clouds, though their position are known(/local-weather/clouds/tile[1]/cloud[...]) - but how to read out and check 100 clouds positions?

I would suggest to do a loop over the cloud properties executed every 3 seconds or so. That shouldn't be too costly as it is just simple math, isn't it?

**Comment by legoboyvdlp** [comment]
@HHS81 may I adapt this in the future for the IDG-A32X?

**Comment by HHS81** [comment]
@legoboyvdlp 
Yes, it is under GNU GPL, so no problem. But the code is more a hack, it only detects the vertical position of cloud layers. But as the Weather radar on the Dornier 328 shows, it should be possible to detect them horizontal as way.... But that's beyond my knowledge.

**Comment by HHS81** [comment]
I'm currently stuck with an more or less unstable computer. Maby in one or two months I can afford a new and better one. Until then overall work is progressing slowly, very slowly.... :disappointed:

**Comment by hbeni** [comment]
@HHS81 if you don't mind i would want to take on this issue next, so we can close it and have the new effects. This would also enable us to have the stall vane icing and testing with just a small effort on your side (modelling the blend file, but the .ac already works) and my side (integrating the stall icing code once pitot icing was merged).

> I also made the fdm react to drag and loss of lift due to ice. But it needs some tuning, the effects seems to small.
> What's missing is the effect of prop icing and increase of weight by ice.

Where can i find the code?
Where can i find some hints on the effects strength?

**Comment by HHS81** [comment]
increase og weight by ice added and working

**Comment by HHS81** [comment]
It seems I can`t alter thrust in runtime, so I leave this out. 
I added the "Aircraft clean" button, and made the dirt map multi, so we could use in a near future also dirt.

**Comment by HHS81** [comment]
@hbeni please have a test

**Comment by hbeni** [comment]
Will have. 
What about a workaround via the power calculation for bhp? It should have the same outcome (engine output less on same rpm)

**Comment by hbeni** [comment]
> please have a test

Basic effects look still good.

@HHS81 do you have any METAR string (or even full command line) you test with so we can get similar results?

**Comment by HHS81** [comment]
@hbeni 
just try the CAT III weather, that was my test setting

**Comment by hbeni** [comment]
Did tests and the effect seems plausible to me. But be aware i have no experience in flying into icing conditions, nor i'm a real pilot.

Also, i did add an effect that engine cooling is reduced with nose ice build up: The Air inlets are iced too, so less air could pass through to cool the engine. This could lead to worsened CHT numbers. I hoped, that the effect is neither too strong nor too weak, but i was able to land in CAT-IIIb weather at zero degrees without killing the engine (cowl flaps opened, throttle reduced and full rich). We can tune that easily if we get some first hand numbers. Meanwhile the effect is there and has to be monitored.

**Comment by hbeni** [comment]
@HHS81 any results yet with the workaround?
From my point of view we might merge it already and could leave the issue open for the drag thing

**Comment by hbeni** [comment]
Another idea would be to just increase drag a little as we did with doors.
I think this would even better separate.

I will try to add this, together with some comments in the trust config.

**Comment by HHS81** [comment]
No, this isn't realistic. There is already drag of the ice simulated. But not the effect of ice on the prop - it does create drag on the prop influencing rpm. 
I already asked on the jsbsim-list and waiting for an answer...

**Comment by hbeni** [comment]
Ah great, should we nonetheless merge the already achieved stuff since utitooks good already?

**Comment by HHS81** [comment]
hmmm...the workaround with the prower calculation is not correct. I think I have a better way by altering secretly altering the prop speed....

**Comment by HHS81** [comment]
btw. we should also consider the real effects of iece on static ports etc. as described by legoboy in the c172p project

**Comment by legoboyvdlp** [comment]
There is a bug that needs fixed in core for that - James said he would do it but is rather busy at the minute it seems.

**Comment by HHS81** [comment]
@hbeni  I added the effect of ice to the prop, and some other things. Please test, it should be much more realistic.
